database/buildSchema/43_views_functions_triggers.sql

```diff
@@ -1683,78 +1681,48 @@ BEGIN
             WHERE id = NEW.application_id;
     END IF;
 
-    -- Get list of reviewer_ids to update for 
-    IF TG_OP = 'UPDATE' THEN
-        reviewer_ids = ARRAY(
-            SELECT reviewer_id FROM review_assignment
-            WHERE application_id = NEW.application_id
-        );
-    ELSE
-        reviewer_ids = ARRAY[NEW.reviewer_id];
-    END IF;
-
-    FOREACH rev_id IN ARRAY reviewer_ids LOOP
-        -- Remove existing
-        UPDATE public.application_reviewer_action
-            SET reviewer_action = NULL
-            WHERE application_id = NEW.application_id
-            AND user_id = rev_id;
-
-        -- Insert new
-        WITH reviewer_action AS (
-            SELECT reviewer_action
-                FROM single_application_detail(NEW.application_id, rev_id)
-            ) 
-        INSERT INTO public.application_reviewer_action
-            (user_id, application_id, reviewer_action)  
-        VALUES(
-            rev_id,
-            NEW.application_id,
-            (SELECT reviewer_action FROM reviewer_action)
-         )
-         ON CONFLICT (user_id, application_id)
-         DO UPDATE
-            SET reviewer_action = (SELECT reviewer_action FROM reviewer_action);
-    END LOOP;
-    
-    
-    -- ASSIGNERS
+    -- REVIEWER/ASSIGNER actions
+    -- When updating review_assignments, we re-calculate for every
+    -- reviewer/assigner associated with this application.
     IF TG_OP = 'UPDATE' THEN
-        assigner_ids = ARRAY(
-            SELECT assigner_id FROM review_assignment
-            WHERE application_id = NEW.application_id
-            AND assigner_id IS NOT NULL
-        );
+        user_ids = ARRAY(
+            -- Creates an array of distinct user ids for all assigners/reviewers
+            -- who can interact with this application
+            SELECT DISTINCT UNNEST(ARRAY[reviewer_id, raaj.assigner_id])
+            FROM review_assignment ra
+            LEFT JOIN review_assignment_assigner_join raaj
+            ON ra.id = raaj.review_assignment_id
+            WHERE application_id = NEW.application_id);
     ELSE
-        CASE WHEN NEW.assigner_id IS NULL THEN
-                assigner_ids = ARRAY[];
-            ELSE
-                assigner_ids = ARRAY[NEW.assigner_id];
-            END CASE;
+    -- But when inserting, just update for the new reviewer
+        user_ids = ARRAY[NEW.reviewer_id];
```

Comment:
This bit doesn't quite work, I think: We also need to get the assigners linked with the new review_id, but because they're created with an Action, we have no guarantee that they'll be created by the time this runs.

Any suggestions?

Reply:
Okay, @andreievg , I think I've got it going properly now. Added an additional (very similar) function that runs on review_assignment_assigner_join INSERTS (only). 

So the triggers are:
- on review_assignment INSERT: run original function (in principle gets assigners from assigner_join too, but in practice those records probably aren't created yet, so will only be for reviewer action). Runs once for *each* insert.
- on review assignment UPDATE: run same function, this time will also update for assigner_joins -- single update re-calculates for *all* related reviewers/assigners
- on review_assignment_assigner_join INSERT: run slightly different function, but does the same as the above but just for the assigner. No need to run on UPDATE as the review_assignment update takes care of it (plus these don't really get updated as a matter of course)

It's somewhat inefficient -- ideally we'd wait for all relevant records to be inserted and then run a function once, but this is not really possible without moving this functionality out into the core Actions, but this would be quite a bit more complicated at this point. We'll see how it goes, speed wise, on the demo site.

Reply:
Thanks Carl, this looks pretty solid, I think it's ok to approve and get it deployed on demo server for performance testing.

I do think that there will be quite a lot of excessive updates, and potential slow down in all operations that have the added trigger, i did a quick check, to see how many assigners are linked to each review_assignment:

```sql
select number_of_assgners, count(*) from 
(select review_assignment.id, count(raj.assigner_id) number_of_assgners from review_assignment join review_assignment_assigner_join raj on raj.review_assignment_id = review_assignment.id group by 1) 
group by 1
```

 number_of_assgners | count
--------------------+-------
                 11 |   976
                  9 |   350
                 15 |   746
                  5 |    12
                  4 |     9
                 10 |    68
                  6 |  6273
                 14 |  6929
                 13 |  8571
                 16 |     9
                  7 |  2319
                 12 | 17188
                  8 |    66

This suggest that majoriy have at least 14 assigners.

I think a queue would be great, doesn't need to be core action queue, can be separate. But yeah lets wait to see the results, thanks for the massive effort